src/psyclone/psyir/nodes/call.py

```diff
@@ -593,3 +610,189 @@ def _location_txt(node):
+        supplied routine.
+
+        :return: None if no match was found, otherwise list of integers
+            referring to matching arguments.
+        :rtype: None|List[int]
+        '''
+
+        # Create a copy of the list of actual arguments to the routine.
+        # Once an argument has been successfully matched, set it to 'None'
+        routine_argument_list: List[DataSymbol] = (
+            routine.symbol_table.argument_list[:]
+        )
+
+        if len(self.arguments) > len(routine.symbol_table.argument_list):
+            raise CallMatchingArgumentsNotFound(
+                f"More arguments in call ('{self.debug_string()}')"
+                f" than callee (routine '{routine.name}')"
+            )
+
+        # Iterate over all arguments to the call
+        ret_arg_idx_list = []
+        for call_arg_idx, call_arg in enumerate(self.arguments):
+            call_arg_idx: int
+            call_arg: DataSymbol
+
+            # If the associated name is None, it's a positional argument
+            # => Just return the index if the types match
+            if self.argument_names[call_arg_idx] is None:
+                routine_arg = routine_argument_list[call_arg_idx]
+                routine_arg: DataSymbol
+
+                self._check_argument_type_matches(call_arg, routine_arg)
+
+                ret_arg_idx_list.append(call_arg_idx)
+                routine_argument_list[call_arg_idx] = None
+                continue
+
+            #
+            # Next, we handle all named arguments
+            #
+            arg_name = self.argument_names[call_arg_idx]
+            routine_arg_idx = None
+
+            for routine_arg_idx, routine_arg in enumerate(
+                routine_argument_list
+            ):
+                routine_arg: DataSymbol
+
+                # Check if argument was already processed
+                if routine_arg is None:
+                    continue
+
+                if arg_name == routine_arg.name:
+                    self._check_argument_type_matches(call_arg, routine_arg)
+                    ret_arg_idx_list.append(routine_arg_idx)
+                    break
+
+            else:
+                # It doesn't match => Raise exception
+                raise CallMatchingArgumentsNotFound(
+                    f"Named argument '{arg_name}' not found"
+                )
+
+            routine_argument_list[routine_arg_idx] = None
+
+        #
+        # Finally, we check if all left-over arguments are optional arguments
+        #
+        for routine_arg in routine_argument_list:
+            routine_arg: DataSymbol
+
+            if routine_arg is None:
+                continue
+
+            # TODO #759: Optional keyword is not yet supported in psyir.
+            # Hence, we use a simple string match.
+            if ", OPTIONAL" not in str(routine_arg.datatype):
+                raise CallMatchingArgumentsNotFound(
+                    f"Argument '{routine_arg.name}' in subroutine"
```

Comment:
This error message and the comment above need a bit of rewording now. e.g. the comment should say something like "Any arguments that haven't been matched must be OPTIONAL otherwise we have an error."
And the error message:
"Argument .... in subroutine .... does not match any in the call '{self.debug_string()}' and is not OPTIONAL.

Reply:
OK.

Reply:
You've missed the comment but again, I can do this.

Reply:
Actually, looking at it in context, this is already covered by an earlier comment so it's fine.